llvm/lib/Target/AMDGPU/SIProgramInfo.h

```diff
@@ -56,44 +59,52 @@ struct SIProgramInfo {
     uint32_t LdsSize = 0;
     uint32_t EXCPEnable = 0;
 
-    uint64_t ComputePGMRSrc3GFX90A = 0;
+    const MCExpr *ComputePGMRSrc3GFX90A = nullptr;
 
-    uint32_t NumVGPR = 0;
-    uint32_t NumArchVGPR = 0;
-    uint32_t NumAccVGPR = 0;
-    uint32_t AccumOffset = 0;
+    const MCExpr *NumVGPR = nullptr;
+    const MCExpr *NumArchVGPR = nullptr;
+    const MCExpr *NumAccVGPR = nullptr;
+    const MCExpr *AccumOffset = nullptr;
     uint32_t TgSplit = 0;
-    uint32_t NumSGPR = 0;
+    const MCExpr *NumSGPR = nullptr;
     unsigned SGPRSpill = 0;
     unsigned VGPRSpill = 0;
     uint32_t LDSSize = 0;
-    bool FlatUsed = false;
+    const MCExpr *FlatUsed = nullptr;
 
     // Number of SGPRs that meets number of waves per execution unit request.
-    uint32_t NumSGPRsForWavesPerEU = 0;
+    const MCExpr *NumSGPRsForWavesPerEU = nullptr;
 
     // Number of VGPRs that meets number of waves per execution unit request.
-    uint32_t NumVGPRsForWavesPerEU = 0;
+    const MCExpr *NumVGPRsForWavesPerEU = nullptr;
 
     // Final occupancy.
-    uint32_t Occupancy = 0;
+    const MCExpr *Occupancy = nullptr;
 
     // Whether there is recursion, dynamic allocas, indirect calls or some other
     // reason there may be statically unknown stack usage.
-    bool DynamicCallStack = false;
+    const MCExpr *DynamicCallStack = nullptr;
 
     // Bonus information for debugging.
-    bool VCCUsed = false;
+    const MCExpr *VCCUsed = nullptr;
 
     SIProgramInfo() = default;
```

Comment:
Should the ctor call `reset()` to enforce that both a reset and a freshly constructed `SIProgramInfo` always have the same values?

Reply:
If this is preferred I can do so (or alternatively, move the reset functionality into the constructor altogether).

I did consider having the reset functionality in `SIProgramInfo`'s constructor but instead I've created an explicit reset as there is a cost involved with allocating a `MCConstantExpr`. If the constructor would have a reset (or reset-like functionality) and a subsequent reset is called on it, the `MCContext` may have allocated unused memory that cannot be deallocated.

Let me know if there's a preference either way.

Reply:
I am just worried about a default changing in the var's declaration and not being updated in `reset()`. I guess it's not worth to slow down execution for this unlikely possibility. Can you just add a comment that mentions all values are duplicated in `reset()` ?

Reply:
I've added a comment on the `reset` hoping to explain this, let me know whether what I have is sufficient.